Sources/Datadog/Core/System/Time/ServerDateProvider.swift

```diff
@@ -10,21 +10,41 @@ import Kronos
 /// Abstract the monotonic clock synchronized with the server using NTP.
 internal protocol ServerDateProvider {
     /// Start the clock synchronisation with NTP server.
-    /// Calls the `completion` by passing it the server time when the synchronization succeeds or`nil` if it fails.
-    func synchronize(with ntpPool: String, completion: @escaping (Date?) -> Void)
-    /// Returns the server time or `nil` if not yet determined.
-    /// This time gets more precise while synchronization is pending.
-    func currentDate() -> Date?
+    /// Calls the `completion` by passing it the server time offset when the synchronization succeeds or`nil` if it fails.
+    func synchronize(with pool: String, completion: @escaping (TimeInterval?) -> Void)
+    /// Returns the server time offset or `nil` if not yet determined.
+    /// This offset gets more precise while synchronization is pending.
+    var offset: TimeInterval? { get }
 }
 
 internal class NTPServerDateProvider: ServerDateProvider {
-    func synchronize(with ntpPool: String, completion: @escaping (Date?) -> Void) {
-        Clock.sync(from: ntpPool, completion: { serverTime, _ in
-            completion(serverTime)
-        })
+    /// Server offset publisher.
+    private let publisher: ValuePublisher<TimeInterval?> = ValuePublisher(initialValue: nil)
+
+    /// Returns the server time offset or `nil` if not yet determined.
+    /// This offset gets more precise while synchronization is pending.
+    var offset: TimeInterval? {
+        return publisher.currentValue
     }
 
-    func currentDate() -> Date? {
-        return Clock.now
+    func synchronize(with pool: String, completion: @escaping (TimeInterval?) -> Void) {
+        // Kronos only notifies for the first and last samples.
+        // In case, the last sample does not return an offset, we calculate the offset
+        // from `Clock.now`.
+        Clock.sync(
+            from: pool,
+            first: { _, offset in
+                self.publisher.publishAsync(offset)
+            },
+            completion: { now, offset in
+                if let offset = offset {
+                    self.publisher.publishAsync(offset)
+                } else if let now = now {
+                    self.publisher.publishAsync(now.timeIntervalSinceNow)
```

Comment:
Is it possible to get `now` but not `offset`, or this is just a sanity fallback? In any case, it would be good to have inline comment on this, otherwise it seems unclear.

Reply:
Yes, my comment above is misplaced. This is indeed a fallback in case the last sample returns a `nil` offset. 
> the `now` parameter is `Clock.now`.

Reply:
I've updated my comment, LMKWYT!